### PR TITLE
feat(tribute-passes): implement cont_to_yield_bubbling pass for native backend

### PR DESCRIPTION
**Pull request by Kroisse**
## Summary

- Replace `cont_to_libmprompt` on the native backend with a new `cont_to_yield_bubbling` pass that lowers the `cont` dialect to an ADT-based yield-bubbling representation, enabling both the Cranelift and WASM backends to process effects through their existing `adt_to_*` passes without new backend work
- Introduce `YieldResult<A>` as `Done(anyref) | Shift(ShiftInfo)` — a portable, runtime-independent representation that replaces the trampoline and libmprompt strategies
- Add 9 new E2E ability tests covering early return, result arm expressions (identity, constant, transform), state mutations, counter patterns, and three nested abilities (36/36 passing, 2 skipped for unimplemented nested-handler edge cases)

### Key implementation components

- `analysis.rs` — effectful function classification and live-variable analysis
- `call_lower.rs` — expand effectful calls into Done/Shift branches with continuation chaining
- `handler_dispatch.rs` — `cont.handler_dispatch` lowering: `scf.loop` with YieldResult dispatch + `cont.done` body inlining
- `shift_lower.rs` — `cont.shift` → capture live state + build `Continuation` + `ShiftInfo` → return `YieldResult::Shift`
- `truncate.rs` — truncate effectful function bodies at effectful calls
- `wrap_returns.rs` — wrap non-effectful returns in `YieldResult::Done`
- `types.rs` — YieldResult/Continuation ADT type helpers
- `patterns.rs` — `cont.push_prompt` lowering: inline body ops + wire into handler dispatch

### Bug fixes discovered during E2E testing

- `fix(trunk-ir)`: hash `Symbol` by string content instead of `Spur` index to ensure stable HashMap iteration order
- `fix(tribute-passes)`: fix RAUW detection order in `create_resume_function` — RAUW detection now runs after state restoration so live-value mappings are correct for counter patterns
- `fix(tribute-passes)`: inline `cont.done` body in handler dispatch — previously `build_done_branch` ignored the result-arm expression entirely, causing wrong output for identity/constant/arithmetic result arms
- `fix(tribute-passes)`: support nested handlers by using fresh evidence parameters in resume functions and running post-processing loops on generated functions

## Test plan

- [ ] `cargo test -p tribute -- e2e_ability_core` — all 36 tests pass, 2 `#[ignore]` for unimplemented nested-handler edge cases
- [ ] `cargo build` — clean build with no new warnings
- [ ] Verify `cont_to_libmprompt` is no longer invoked in the native pipeline (`src/pipeline.rs`)
- [ ] Verify `cont_to_trampoline` / WASM pipeline is unchanged

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a public cont→yield‑bubbling lowering pass with full pipeline integration, ADT-based YieldResult types, and resume/chain generation.

* **Bug Fixes**
  * Made symbol hashing content‑stable.
  * Improved diagnostics and logging in translation and lowering paths.

* **Tests**
  * Expanded and re‑enabled extensive handler and multi‑ability tests covering nested and edge cases.

* **Chores**
  * Adjusted pipelines and error propagation to surface lowering failures and update native/WASM flows.

### CODE REVIEW

File: crates/tribute-passes/src/cont_to_yield_bubbling/handler_dispatch.rs [hunk 1]
```diff
@@ -0,0 +1,887 @@
...
+        let is_effectful_call = if let Ok(call) = arena_func::Call::from_op(ctx, op) {
+            let callee = call.callee(ctx);
+            effectful_funcs.contains(&callee) && !ctx.op_results(op).is_empty()
+        } else {
+            false
+        };
+
+        let is_resume = arena_cont::Resume::from_op(ctx, op).is_ok();
+        let produces_yr = is_effectful_call || is_resume;
+
```
[the remark this answers is not in the document]
Comment (Kroisse) [reply crates/tribute-passes/src/cont_to_yield_bubbling/handler_dispatch.rs]: Fixed in 688588e3 — removed the `break` so trailing ops after `produces_yr` continue to be cloned.

File: crates/tribute-passes/src/cont_to_yield_bubbling/patterns.rs [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+                ctx,
+                location,
+                [null_op.result(ctx)],
+                t.yield_result,
+                t.yield_result,
+                Symbol::new("Done"),
+            );
+            all_ops.push(done_op.op_ref());
+            done_op.result(ctx)
+        };
```
[the remark this answers is not in the document]
Comment (Kroisse) [reply crates/tribute-passes/src/cont_to_yield_bubbling/patterns.rs]: Not applicable — when `is_yield_result_type` is true (line 350), `_yr_value` is set to `result` and no new ops are pushed. The `all_ops` list contains the inlined body ops, and `all_ops.pop()` correctly gives the last body op (which produces the YieldResult) for `replace_op`. The value flows correctly because the op producing `result` is the same op that becomes the replacement.

Separately, we did fix `region_yields_yr` (used by `UpdateScfIfResultTypePattern`) to check `scf.yield` operand types instead of result types, since `scf.yield` has no results.

File: crates/tribute-passes/src/cont_to_yield_bubbling/handler_dispatch.rs [hunk 3]
```diff
@@ -0,0 +1,886 @@
...
+        hd_ctx.our_tag,
+        hd_ctx.runtime_tag_operand.is_some(),
+        hd_ctx.suspend_arms.len(),
+    );
+
+    let loop_result_ty = if hd_ctx.is_in_effectful_func {
+        hd_ctx.types.yield_result
+    } else {
+        hd_ctx.user_result_ty
+    };
```
[the remark this answers is not in the document]
Comment (Kroisse) [reply crates/tribute-passes/src/cont_to_yield_bubbling/handler_dispatch.rs]: Not applicable — non-effectful handlers handle all effects locally, so the tag-mismatch propagate path is unreachable in practice. The type mismatch is latent but never triggered at runtime.

File: crates/tribute-passes/src/cont_to_yield_bubbling/truncate.rs [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+        let call_result = ctx.op_results(op)[0];
+        let location = ctx.op(op).location;
+
+        // Skip if the call result is consumed by handler_dispatch infrastructure
+        // (scf.loop, adt.variant_new Done). These are already handled by the
+        // handler dispatch loop and should not be unwrapped.
+        let uses = ctx.uses(call_result).to_vec();
+        let consumed_by_handler = uses.iter().any(|u| {
+            scf::Loop::from_op(ctx, u.user).is_ok() || adt::VariantNew::from_op(ctx, u.user).is_ok()
+        });
```
[the remark this answers is not in the document]
Comment (Kroisse) [reply crates/tribute-passes/src/cont_to_yield_bubbling/truncate.rs]: This check runs only inside `unwrap_yr_calls_in_block`, which processes non-effectful functions. In that context, the only `adt.variant_new` ops that consume a YieldResult value are handler dispatch infrastructure (`adt.variant_new Done`). No other VariantNew in a non-effectful function would take a YieldResult as an operand, so the broad check is correct here. Adding a comment to clarify this invariant.

File: crates/tribute-passes/src/cont_to_yield_bubbling/truncate.rs [hunk 5]
```diff
@@ -0,0 +1,714 @@
...
+
+        // Fix 2: Replace all uses of Done result with call_indirect result
+        let done_result = ctx.op_results(op)[0];
+        let call_result = ctx.op_results(call_op)[0];
+        ctx.replace_all_uses(done_result, call_result);
+
+        // Fix 3: Remove the cast and Done ops (they're now dead)
+        ctx.remove_op_from_block(block, op);
+        ctx.remove_op_from_block(block, cast_op);
+
```
[the remark this answers is not in the document]
Comment (Kroisse) [reply crates/tribute-passes/src/cont_to_yield_bubbling/truncate.rs]: The cast is always single-use in this pattern (created by `LowerPushPromptPattern` specifically to feed the Done variant). A use-count guard via `ctx.uses()` does not work here because `remove_op_from_block` does not clean up use chains — the removed Done op still appears as a user, so the guard never allows removal. Leaving the cast unconditionally removed since the pattern match already ensures the exact `call_indirect → cast → Done` chain.

File: crates/tribute-passes/src/cont_to_yield_bubbling/wrap_returns.rs [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+            return false;
+        };
+        if !self.effectful_funcs.contains(&func_name) {
+            return false;
+        }
+
+        // Check if return value is already YieldResult
+        let Some(&value) = ctx.op_operands(op).first() else {
+            return false;
+        };
```
[the remark this answers is not in the document]
Comment (Kroisse) [reply crates/tribute-passes/src/cont_to_yield_bubbling/wrap_returns.rs]: Tribute functions always return exactly 0 or 1 values — multi-value returns are not part of the language design. `func.return` in Tribute IR never has >1 operand, so `.first()` is correct here without an arity guard.
